toolchain/check/type_structure.cpp

```diff
@@ -0,0 +1,455 @@
+          // TODO: The need for `.closing` goes away when other_requirements
+          // does. Are there other places we need to look for symbolics in
+          // FacetTypeInfo at that point? For now we treat it as having a
+          // symbolic at the end of the facet type, so facet types with
+          // other_requirements are chosen with lower priority than those
+          // without.
+          Push(CloseType{.closing = facet_type_info.other_requirements
+                                        ? CloseWithSymbolic
+                                        : CloseWithConcrete});
+          for (const auto& i : facet_type_info.impls_constraints) {
+            PushArgs(GetSpecificArgs(i.specific_id));
+          }
+          break;
+        }
+        case CARBON_KIND(SemIR::TupleType tuple_type): {
+          auto inner_types = context_.type_blocks().Get(tuple_type.elements_id);
+          ++position_;  // Opening of type.
+          if (inner_types.empty()) {
+            structure_.push_back(TypeStructure::Structural::Concrete);
+          } else {
+            structure_.push_back(TypeStructure::Structural::ConcreteOpenParen);
+            Push(CloseType{});
+            for (auto type :
+                 context_.type_blocks().Get(tuple_type.elements_id)) {
+              Push(type);
+            }
+          }
+          break;
+        }
+        case CARBON_KIND(SemIR::StructType struct_type): {
+          auto fields =
+              context_.struct_type_fields().Get(struct_type.fields_id);
+          ++position_;  // Opening of type.
+          if (fields.empty()) {
+            structure_.push_back(TypeStructure::Structural::Concrete);
+          } else {
+            structure_.push_back(TypeStructure::Structural::ConcreteOpenParen);
+            Push(CloseType{});
+            for (const auto& field : fields) {
+              Push(field.type_id);
+            }
+          }
+          break;
+        }
+        default:
+          CARBON_FATAL("Unhandled type instruction {0}", inst_id);
+      }
+    }
+  }
+
+  enum CloseTypeClosing {
+    CloseOnly,
+    CloseWithConcrete,
+    CloseWithSymbolic,
+  };
+  struct CloseType {
+    CloseTypeClosing closing = CloseOnly;
+  };
+  struct SymbolicType {};
+  struct NonTypeValue {};
+
+  using WorkItem = std::variant<SemIR::TypeId, SymbolicType, NonTypeValue,
+                                SemIR::SpecificInterface, CloseType>;
+
+  auto TryGetInstIdAsTypeId(SemIR::InstId inst_id) const
+      -> std::variant<SemIR::TypeId, SymbolicType> {
+    if (auto facet_value =
+            context_.insts().TryGetAs<SemIR::FacetValue>(inst_id)) {
+      inst_id = facet_value->type_inst_id;
+    }
+
+    auto type_id_of_inst_id = context_.insts().Get(inst_id).type_id();
+    if (context_.types().Is<SemIR::FacetType>(type_id_of_inst_id)) {
+      return SymbolicType();
+    }
+    if (type_id_of_inst_id != SemIR::TypeType::SingletonTypeId) {
+      return SemIR::TypeId::None;
+    }
```

Comment:
A comment explaining why it's correct to treat non-type facets as always being symbolic would be useful. (I think this might be right, but I'm not certain.)

Reply:
Done